### PR TITLE
Add `CONFIG GET` and `CONFIG SET` commands.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Features:
* `CONFIG GET` https://redis.io/commands/config-get/
* `CONFIG SET` https://redis.io/commands/config-set/

### CODE REVIEW

File: java/client/src/main/java/glide/api/RedisClusterClient.java [hunk 1]
```diff
@@ -118,4 +121,36 @@ public CompletableFuture<ClusterValue<String>> info(
         return commandManager.submitNewCommand(
                 Info, options.toArgs(), route, response -> ClusterValue.of(handleObjectResponse(response)));
     }
+
+    /** {@inheritDoc} The command will be sent to a random node. */
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/RedisClusterClient.java:125]: image

File: java/client/src/main/java/glide/Utils.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/** Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0 */
+package glide;
+
+import java.util.Map;
+import java.util.stream.Stream;
+
+public class Utils {
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/Utils.java:7]: `ArrayUtils`?

File: java/client/src/main/java/glide/Utils.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    // TODO https://github.com/aws/glide-for-redis/pull/955/files#r1488732655
+    //   Rebase & merge after #959 and #955
+    public static String[] convertMapToArgArray(Map<String, String> args) {
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/Utils.java:10]:
```suggestion
    public static String[] convertMapToArray(Map<String, String> args) {
```

File: java/client/src/main/java/glide/api/commands/ServerManagementBaseCommands.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+public interface ServerManagementBaseCommands {
+
+    /**
+     * Read the configuration parameters of a running Redis server.
+     *
+     * @see <a href="https://redis.io/commands/config-get/">redis.io</a> for details.
+     * @param parameters A list of configuration parameter names to retrieve values for.
+     * @return A map of values corresponding to the configuration parameters.
+     */
+    CompletableFuture<Map<String, String>> configGet(String... parameters);
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ServerManagementBaseCommands.java:16]:
```suggestion
    CompletableFuture<Map<String, String>> configGet(String[] parameters);
```

File: java/client/src/main/java/glide/api/commands/ServerManagementBaseCommands.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.concurrent.CompletableFuture;
+
+public interface ServerManagementBaseCommands {
+
+    /**
+     * Read the configuration parameters of a running Redis server.
+     *
+     * @see <a href="https://redis.io/commands/config-get/">redis.io</a> for details.
+     * @param parameters A list of configuration parameter names to retrieve values for.
+     * @return A map of values corresponding to the configuration parameters.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ServerManagementBaseCommands.java:14]:
```suggestion
     * @return A <code>Map</code> of values corresponding to the configuration parameters.
```

File: java/client/src/main/java/glide/api/commands/ServerManagementBaseCommands.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+     */
+    CompletableFuture<Map<String, String>> configGet(String... parameters);
+
+    /**
+     * Set configuration parameters to the specified values.
+     *
+     * @see <a href="https://redis.io/commands/config-set/">redis.io</a> for details.
+     * @param parameters A map consisting of configuration parameters and their respective values to
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ServerManagementBaseCommands.java:22]:
```suggestion
     * @param parameters A <code>Map</code> consisting of configuration parameters and their respective values to
```

File: java/client/src/main/java/glide/api/commands/ServerManagementClusterCommands.java [hunk 7]
```diff
@@ -70,4 +71,30 @@ public interface ServerManagementClusterCommands {
...
+     *     When specifying a route other than a single node, it returns a dictionary where each
+     *     address is the key and its corresponding node response is the value.
+     */
+    CompletableFuture<ClusterValue<Map<String, String>>> configGet(Route route, String... parameters);
+
+    /**
+     * Set configuration parameters to the specified values.
+     *
+     * @see <a href="https://redis.io/commands/config-set/">redis.io</a> for details.
+     * @param parameters A map consisting of configuration parameters and their respective values to
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ServerManagementClusterCommands.java:92]:
```suggestion
     * @param parameters A <code>Map</code> consisting of configuration parameters and their respective values to
```

File: java/client/src/main/java/glide/api/commands/ServerManagementClusterCommands.java [hunk 8]
```diff
@@ -70,4 +71,30 @@ public interface ServerManagementClusterCommands {
...
     CompletableFuture<ClusterValue<String>> info(InfoOptions options, Route route);
+
+    /**
+     * Read the configuration parameters of a running Redis server.
+     *
+     * @see <a href="https://redis.io/commands/config-get/">redis.io</a> for details.
+     * @param route Routing configuration for the command. Client will route the command to the nodes
+     *     defined.
+     * @param parameters A list of configuration parameter names to retrieve values for.
+     * @return A map of values corresponding to the configuration parameters.<br>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ServerManagementClusterCommands.java:82]:
```suggestion
     * @return A <code>Map</code> of values corresponding to the configuration parameters.<br>
```

File: java/client/src/main/java/glide/api/commands/ServerManagementClusterCommands.java [hunk 9]
```diff
@@ -70,4 +71,30 @@ public interface ServerManagementClusterCommands {
...
+     *     When specifying a route other than a single node, it returns a dictionary where each
+     *     address is the key and its corresponding node response is the value.
+     */
+    CompletableFuture<ClusterValue<Map<String, String>>> configGet(Route route, String... parameters);
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ServerManagementClusterCommands.java:86]:
```suggestion
    CompletableFuture<ClusterValue<Map<String, String>>> configGet(String[] parameters, Route route);
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 10]
```diff
@@ -114,6 +118,36 @@ public T info(InfoOptions options) {
         return getThis();
     }
 
+    /**
+     * Read the configuration parameters of a running Redis server.
+     *
+     * @see <a href="https://redis.io/commands/config-get/">redis.io</a> for details.
+     * @param parameters A list of configuration parameter names to retrieve values for.
+     * @return A map of values corresponding to the configuration parameters.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:126]:
```suggestion
     * @return A <code>Map</code> of values corresponding to the configuration parameters.
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 11]
```diff
@@ -114,6 +118,36 @@ public T info(InfoOptions options) {
...
+     */
+    public T configGet(String... parameters) {
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:128]:
```suggestion
    public T configGet(String[] parameters) {
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 12]
```diff
@@ -114,6 +118,36 @@ public T info(InfoOptions options) {
...
+
+        protobufTransaction.addCommands(buildCommand(ConfigGet, commandArgs));
+        return getThis();
+    }
+
+    /**
+     * Set configuration parameters to the specified values.
+     *
+     * @see <a href="https://redis.io/commands/config-set/">redis.io</a> for details.
+     * @param parameters A map consisting of configuration parameters and their respective values to
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:139]:
```suggestion
     * @param parameters A <code>Map</code> consisting of configuration parameters and their respective values to
```

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 13]
```diff
@@ -251,4 +254,51 @@ public void info_with_route_with_infoOptions_returns_string() {
         assertEquals("addr1 result", clusterMap.get("addr1"));
         assertEquals("addr2 result", clusterMap.get("addr2"));
     }
+
+    // TODO copy tests from RedisClientTest which have @inheritDoc for coverage
+
+    @Test
+    @SneakyThrows
+    // test checks that even a map returned as a single value when single node route is used
+    public void config_get_with_single_node_route_returns_single_value() {
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:263]:
```suggestion
    public void configGet_with_single_node_route_returns_single_value() {
```

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 14]
```diff
@@ -251,4 +254,51 @@ public void info_with_route_with_infoOptions_returns_string() {
...
+        var client = new TestClient(commandManager, data);
+
+        var value = client.configGet(RANDOM, TEST_ARGS).get();
+        assertAll(
+                () -> assertTrue(value.hasSingleData()), () -> assertEquals(data, value.getSingleValue()));
+    }
+
+    @Test
+    @SneakyThrows
+    public void config_get_with_multi_node_route_returns_multi_value() {
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:276]:
```suggestion
    public void configGet_with_multi_node_route_returns_multi_value() {
```

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 15]
```diff
@@ -251,4 +254,51 @@ public void info_with_route_with_infoOptions_returns_string() {
...
+        var client = new TestClient(commandManager, data);
+
+        var value = client.configGet(ALL_NODES, TEST_ARGS).get();
+        assertAll(
+                () -> assertTrue(value.hasMultiData()), () -> assertEquals(data, value.getMultiValue()));
+    }
+
+    @SneakyThrows
+    @Test
+    public void config_set_with_route_returns_success() {
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:289]:
```suggestion
    public void configSet_with_route_returns_success() {
```

File: java/client/src/test/java/glide/api/models/ClusterTransactionTests.java [hunk 16]
```diff
@@ -57,6 +60,16 @@ public void transaction_builds_protobuf_request() {
                         Info,
                         ArgsArray.newBuilder().addArgs(InfoOptions.Section.EVERYTHING.toString()).build()));
 
+        transaction.configGet("Foster", "The", "Fluffy", "Dogster");
+        results.add(Pair.of(ConfigGet, transaction.buildArgs("Foster", "The", "Fluffy", "Dogster")));
+
+        var configSetMap = new LinkedHashMap<String, String>();
+        configSetMap.put("Foster", "The");
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/models/ClusterTransactionTests.java:67]: 🐶

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 17]
```diff
@@ -185,7 +230,93 @@ public void info_with_routing_and_options() {
...
-            assertTrue(data.getSingleValue().toLowerCase().contains("# " + section.toLowerCase()), "Section " + section + " is missing");
+            assertTrue(
+                    data.getSingleValue().toLowerCase().contains("# " + section.toLowerCase()),
+                    "Section " + section + " is missing");
         }
     }
+
+    @Test
+    @SneakyThrows
+    public void config_get_with_no_args_returns_error() {
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:241]:
```suggestion
    public void configGet_with_no_args_returns_error() {
```

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 18]
```diff
@@ -185,7 +230,93 @@ public void info_with_routing_and_options() {
...
+        var exception = assertThrows(ExecutionException.class, () -> clusterClient.configGet().get());
+        assertTrue(exception.getCause() instanceof RedisException);
+        assertTrue(exception.getCause().getMessage().contains("wrong number of arguments"));
+    }
+
+    @Test
+    @SneakyThrows
+    public void config_get_with_wildcard() {
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:249]:
```suggestion
    public void configGet_with_wildcard() {
```

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 19]
```diff
@@ -185,7 +230,93 @@ public void info_with_routing_and_options() {
...
+        var data = clusterClient.configGet("*file").get();
+        assertTrue(data.size() > 5);
+        assertTrue(data.containsKey("pidfile"));
+        assertTrue(data.containsKey("logfile"));
+    }
+
+    @Test
+    @SneakyThrows
+    public void config_get_with_multiple_params() {
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:258]:
```suggestion
    public void configGet_with_multiple_params() {
```

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 20]
```diff
@@ -185,7 +230,93 @@ public void info_with_routing_and_options() {
...
+        assumeTrue(REDIS_VERSION.feature() >= 7, "This feature added in redis 7");
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:259]: I didn't know about `assume`

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 21]
```diff
@@ -185,7 +230,93 @@ public void info_with_routing_and_options() {
...
+        var data = clusterClient.configGet("pidfile", "logfile").get();
+        assertAll(
+                () -> assertEquals(2, data.size()),
+                () -> assertTrue(data.containsKey("pidfile")),
+                () -> assertTrue(data.containsKey("logfile")));
+    }
+
+    @Test
+    @SneakyThrows
+    public void config_get_with_wildcard_and_single_node_route() {
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:269]:
```suggestion
    public void configGet_with_wildcard_and_single_node_route() {
```

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 22]
```diff
@@ -185,7 +230,93 @@ public void info_with_routing_and_options() {
...
+        assertTrue(data.hasSingleData());
+        assertAll(
+                () -> assertTrue(data.getSingleValue().size() > 5),
+                () -> assertTrue(data.getSingleValue().containsKey("pidfile")),
+                () -> assertTrue(data.getSingleValue().containsKey("logfile")));
+    }
+
+    @Test
+    @SneakyThrows
+    public void config_get_with_wildcard_and_multi_node_route() {
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:280]: same

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 23]
```diff
@@ -185,7 +230,93 @@ public void info_with_routing_and_options() {
...
+                data.getMultiValue().get(data.getMultiValue().keySet().toArray(String[]::new)[0]);
+        assertAll(
+                () -> assertTrue(config.size() > 5),
+                () -> assertTrue(config.containsKey("pidfile")),
+                () -> assertTrue(config.containsKey("logfile")));
+    }
+
+    @Test
+    @SneakyThrows
+    public void config_set_a_parameter() {
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:294]:
```suggestion
    public void configSet_a_parameter() {
```

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 24]
```diff
@@ -185,7 +230,93 @@ public void info_with_routing_and_options() {
...
+        var newValue = clusterClient.configGet("maxclients").get();
+        assertEquals("42", newValue.get("maxclients"));
+
+        response = clusterClient.configSet(Map.of("maxclients", oldValue)).get();
+        assertEquals(OK, response);
+    }
+
+    @Test
+    @SneakyThrows
+    public void config_set_a_parameter_with_routing() {
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:308]:
```suggestion
    public void configSet_a_parameter_with_routing() {
```

File: java/integTest/src/test/java/glide/standalone/CommandTests.java [hunk 25]
```diff
@@ -99,4 +106,62 @@ public void info_with_everything_option() {
             assertTrue(data.contains("# " + section), "Section " + section + " is missing");
         }
     }
+
+    @Test
+    @SneakyThrows
+    public void config_get_with_no_args_returns_error() {
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/standalone/CommandTests.java:112]: Please fix to configGet and configSet throughout the tests.

File: java/integTest/src/test/java/glide/standalone/CommandTests.java [hunk 26]
```diff
@@ -99,4 +106,62 @@ public void info_with_everything_option() {
...
+        var exception = assertThrows(ExecutionException.class, () -> regularClient.configGet().get());
+        assertTrue(exception.getCause() instanceof RedisException);
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/standalone/CommandTests.java:114]: RedisException or RequestException?

File: java/integTest/src/test/java/glide/standalone/CommandTests.java [hunk 27]
```diff
@@ -99,4 +106,62 @@ public void info_with_everything_option() {
...
+    }
+
+    @Test
+    @SneakyThrows
+    public void config_set_with_unknown_parameter_returns_error() {
+        var exception =
+                assertThrows(
+                        ExecutionException.class,
+                        () -> regularClient.configSet(Map.of("puppy", "happy")).get());
+        assertTrue(exception.getCause() instanceof RedisException);
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/standalone/CommandTests.java:145]: RedisException or RequestException?

**Review by acarbonetto — APPROVED** [review]
A couple of minor things.  And we should use String[] not varargs

**Pull request merged**